### DISCUSSION TITLE
Build with docking + implot by default

### DISCUSSION THREAD

**Discussion by yamen** [Gnimuc/CImGui.jl#43] [General]
I'm doing some work that needed the latest DearImGui (1.81) plus the docking branch plus the latest implot. I'd like to suggest that CImGui_jll should:

- Default to the docking branch, which is stable and kept up to date with main
- Default to including the latest relevant implot build, as this is recommended to be in the same DLL for compatibility and performance

I've validated a binary builder that does this (updated script here https://github.com/yamen/CImGui_jll-builder and generated to https://github.com/yamen/CImGui_jll.jl), regenerated CImGui.jll against this including manually making relevant source changes (https://github.com/yamen/CImGui.jl).

I've also created versions of ImPlot.jl that work with this new binary and all is looking good!

Note I'm not suggesting raising a PR against these files yet (they need a little more validation). I wanted to first discuss whether or not this is the right path to go down. It would help in minimising lag between imgui/implot updates and the ability to integrate them into Julia projects. CImGui.jl and ImPlot.jl can evolve wrappers slower, but the embedded LibCImGui and LibCImPlot can move fast and users can directly use these with confidence in API stability.

**Comment by Gnimuc** [comment]
Thanks for looking into this! This is definitely the right path to go. The only thing that I concerned about is that after switching to the docker branch, is there anything that needs to be changed on the backend?

**Reply by yamen** [reply]
I don’t believe so, you can compare generated outputs and see that it’s 100% backwards compatible. Turning on docking in your app is also an explicit flag that needs to be set. Id also suggest that if we allow the jll builds to progress ahead of the Julia wrappers, then users can choose to directly invoke the C library and take the risk/reward themselves. Version stability can be added in Julia wrappers if desired. This gives us best of both worlds.

**Reply by yamen** [reply]
One other comment since I had to do a regen today - I believe the new branch includes Viewports, which means more references to ImGuiWindow / ImViewport in the _common.jl file. Some of these are forward references that don't work, so it increases the number of manual changes that need to be made to the generated files. Takes a few minutes after generation to go through and mod these.

I believe you were working on Clang.jl to improve this? Not sure if that's made it in yet (is that what #42 is for?). Anyway worth knowing implication either way.

**Reply by sonoro1234** [reply]
> 
> 
> Thanks for looking into this! This is definitely the right path to go. The only thing that I concerned about is that after switching to the docker branch, is there anything that needs to be changed on the backend?

You can see the only changes needed for docking and viewport in https://github.com/cimgui/cimgui/blob/master/backend_test/main.c#L87 under `IMGUI_HAS_DOCK`

**Comment by Gnimuc** [comment]
The old Clang.jl generator could generate something wrong for the bitfields structs in cimgui v1.81. #42 is the fix.

**Reply by yamen** [reply]
Want to try it against 1.81 docking plus implot? You can generate the jll from the repo I linked. Note this references my own build of cimplot as I added implot_internal.h to the library. Separately working through a PR to get that merged to cimplot. 

Probably worth reviewing this also to see if new generator will work with it? https://github.com/yamen/ImPlot.jl/blob/master/gen/generator.jl

**Comment by Gnimuc** [comment]
I'm also considering creating an org like https://github.com/cimgui for hosting cimgui related Julia packages, so more people could get involved in the development.

**Comment by Gnimuc** [comment]
As ImGui is not recommended to be used as a shared library, I believe it's better to create a `CImGuiPack_jll` that compiles all of the imgui(docking), cimgui(docking), implot, and cimplot into a shared library.

**Reply by yamen** [reply]
@Gnimuc Is this something you're willing to do? Would you like me to do it? Personally I'd like to include all the libraries that are currently being wrapped by https://github.com/cimgui/ to make it conceptually simplest. That is, when that organisation updates, we can update. Everything downstream (eg CImGui.jl, ImPlot.jl etc) can operate at a slower cadence.

I'm keen to get this done but would like to be aligned with your direction.

@wsphillips thoughts?

**Reply by Gnimuc** [reply]
> Would you like me to do it?

It would be great if you could help with this. This project has a rather low priority on my to-do list.

**Reply by wsphillips** [reply]
>@wsphillips thoughts?

Go for it.

**Comment by Gnimuc** [comment]
FYI, I'm starting working on this.

**Comment by Gnimuc** [comment]
Locked. Let's continue the discussion at #52.
